flink-runtime/src/main/java/org/apache/flink/runtime/security/SecurityConfiguration.java

```diff
@@ -138,25 +141,20 @@ public String getZooKeeperLoginContextName() {
     }
 
     private void validate() {
-        if (!StringUtils.isBlank(keytab)) {
-            // principal is required
-            if (StringUtils.isBlank(principal)) {
-                throw new IllegalConfigurationException(
-                        "Kerberos login configuration is invalid: keytab requires a principal.");
-            }
+        if (StringUtils.isBlank(keytab) != StringUtils.isBlank(principal)) {
+            throw new IllegalConfigurationException(
+                    KERBEROS_CONFIG_ERROR_PREFIX
+                            + "either both keytab and principal must be defined, or neither.");
+        }
 
-            // check the keytab is readable
+        if (!StringUtils.isBlank(keytab)) {
             File keytabFile = new File(keytab);
             if (!keytabFile.exists() || !keytabFile.isFile()) {
                 throw new IllegalConfigurationException(
-                        "Kerberos login configuration is invalid: keytab ["
-                                + keytab
-                                + "] doesn't exist!");
+                        KERBEROS_CONFIG_ERROR_PREFIX + "keytab [" + keytab + "] doesn't exist!");
             } else if (!keytabFile.canRead()) {
```

Comment:
FYI `canRead` is not tested because on some OS it's not implemented properly: https://bugs.java.com/bugdatabase/view_bug.do?bug_id=6203387